src/submodules/pathfinding/astar.jl

```diff
@@ -0,0 +1,239 @@
+"""
+    Path{D,T}
+Alias of `MutableLinkedList{NTuple{D,T}}`. Used to represent the path to be
+taken by an agent in a `D` dimensional space.
+"""
+const Path{D,T} = MutableLinkedList{NTuple{D,T}}
+
+struct AStar{D,P,M,T} <: GridPathfinder{D,P,M}
+    agent_paths::Dict{Int,Path{D,T}}
+    dims::NTuple{D,T}
+    neighborhood::Vector{CartesianIndex{D}}
+    admissibility::Float64
+    walkable::BitArray{D}
+    cost_metric::CostMetric{D}
+
+    function AStar{D,P,M,T}(
+        agent_paths::Dict,
+        dims::NTuple{D,T},
+        neighborhood::Vector{CartesianIndex{D}},
+        admissibility::Float64,
+        walkable::BitArray{D},
+        cost_metric::CostMetric{D},
+    ) where {D,P,M,T}
+        @assert all(dims .> 0) "Invalid pathfinder dimensions: $(dims)"
+        T <: Integer && @assert size(walkable) == dims "Walkmap must be same dimensions as grid"
+        @assert admissibility >= 0 "Invalid value for admissibility: $admissibility ≱ 0"
+        if cost_metric isa PenaltyMap{D}
+            @assert size(cost_metric.pmap) == size(walkable) "Penaltymap dimensions must be same as walkable map"
+        elseif cost_metric isa DirectDistance{D}
+            if M
+                @assert length(cost_metric.direction_costs) >= D "DirectDistance direction_costs must have as many values as dimensions"
+            else
+                @assert length(cost_metric.direction_costs) >= 1 "DirectDistance direction_costs must have non-zero length"
+            end
+        end
+        new(agent_paths, dims, neighborhood, admissibility, walkable, cost_metric)
+    end
+end
+
+"""
+    Pathfinding.AStar(space::GridSpace{D}; kwargs...)
+    Pathfinding.AStar(space::ContinuousSpace{D}, walkmap::BitArray{D}; kwargs...)
+    Pathfinding.AStar(space::ContinuousSpace{D}, cost_metric::PenaltyMap{D}; kwargs...)
+Enables pathfinding for agents in the provided `space` (which can be a [`GridSpace`](@ref) or
+[`ContinuousSpace`](@ref)) using the A* algorithm. This struct must be passed into any
+pathfinding functions.
+
+For [`ContinuousSpace`](@ref), a walkmap or instance of [`PenaltyMap`](@ref) must be provided
+to specify the level of discretisation of the space.
+
+## Keywords
+- `diagonal_movement = true` specifies if movement can be to diagonal neighbors of a
+  tile, or only orthogonal neighbors. Only available for [`GridSpace`](@ref)
+- `admissibility = 0.0` allows the algorithm to aprroximate paths to speed up pathfinding.
+  A value of `admissibility` allows paths with at most `(1+admissibility)` times the optimal
+  length.
+- `walkable = trues(size(space))` specifies the (un)walkable positions of the
```

Comment:
What's going on here? This is so confusing, why are the call signatures so different between the two spaces??? In one it is called `walkmap` in the other it is called `walkable`. One needs them as a positional argument, the other as a keyword. Why? Make everything keyword, the same keyword, and just add a test inside that only one of the two is given. Set default `walkable` to `nothing` and the default cost_metirci as is (DirectDistance). Then check that either walkable is not nothing, or that the cost metric is PenaltyMap and say "one of these two, and onlky one, must be true". 

Also, the way the docstring is written now, it is confusing whether continuous and discrete spaces are truly the same.

Reply:
I'll work on changing the constructor. There is some naming conflict with `walkable`/`walkmap`, specifically because the struct field is called `walkable` and the function that exposed it in the original API was called `walkmap`. To this end, I think it's better if we keep the name as `walkmap` everywhere

Reply:
Also, wouldn't this remove the ability to have a penaltymap as well as a walkmap? For example, this would make a 2D version of rabbit-fox-hawk impossible. We could just check that their dimensions agree.

Reply:
Okay sorry I had to say "at least one of these two must be true".